### ISSUE TITLE
Build shared libs, add run-export, register in global pinning

### ISSUE THREAD

**Issue by h-vetinari** [conda-forge/cpp-opentelemetry-sdk-feedstock#38]
Hey all

I just noticed that arrow comes with an ability to build against opentelemetry, so it would be nice to have this lib ready for that use-case.

AFAICT from the various feedstocks floating around, this is the main one for the CPP lib, but it:
* build [static](https://github.com/regro/libcfgraph/blob/master/artifacts/cpp-opentelemetry-sdk/conda-forge/linux-64/cpp-opentelemetry-sdk-1.8.1-h367e6c9_0.json#L417-L439) instead of shared libs
* doesn't have a run-export
* is not part of the global [pinning](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/main/recipe/conda_build_config.yaml)

Those things would be the minimum necessary for arrow to build against this feedstock.

**Comment by h-vetinari** [comment]
@lidavidm, any thoughts on this?

**Comment by lidavidm** [comment]
Ah, sorry.

All of these should be fixed, but I haven't had a chance to do much other than merge updates for this.

**Comment by h-vetinari** [comment]
Ok, this was mostly fixed as of #44.

I'd still like to rename the library to something like `libopentelemetry` according to https://github.com/conda-forge/conda-forge.github.io/issues/1073, rather than have _yet another_ library naming pattern (we can keep the old name as a compat wrapper for a while).

But the opentelemetry stuff is distributed over quite a few feedstocks[^1] and I'm not sure if that's conflicting with something else.

Thoughts @lidavidm @wjones127 @xhochy?

[^1]: also notices that the CMake installation files are somewhere in an API package as well and get clobbered (so possibly broken metadata here)

**Comment by lidavidm** [comment]
renaming is a-ok by me.

the clobbering is an unfortunate aspect of: some things want to just use opentelemetry as a producer/API-only, and don't want to pull in the actual implementation bits, only the headers. But the package itself isn't really designed with that in mind. So we're basically building it twice, and so things get clobbered. I'm not sure how best to deal with this.

**Comment by lidavidm** [comment]
Maybe upstream would be open to splitting up the package further.

**Comment by h-vetinari** [comment]
> renaming is a-ok by me.

Cool! Do you think we should rename the API output in parallel too? Looking at the open-telemetry org and its [repos](https://github.com/orgs/open-telemetry/repositories?type=all), we might want to keep the `-cpp` since there are many different languages.

So perhaps:
* `libopentelemetry-cpp` & `libopentelemetry-cpp-api`?

(I'll also note that looking at the [content](https://github.com/regro/libcfgraph/blob/master/artifacts/cpp-opentelemetry-api/conda-forge/linux-64/cpp-opentelemetry-api-1.9.0-hf52228f_0.json#L235-L347) of the package, it's effectively just a header-only package; a similar naming discussion is ongoing for [boost](https://github.com/conda-forge/boost-cpp-feedstock/issues/137) currently, where we might use `libboost-header-only`)

> I'm not sure how best to deal with this.

How about not including the CMake metadata in the api package? I opened a PR: https://github.com/conda-forge/cpp-opentelemetry-api-feedstock/pull/17

**Comment by lidavidm** [comment]
libopentelemetry-cpp/-cpp-api sounds good to me.

The -api scheme is because that's what the package itself calls the header-only variant, not sure which is preferable but I don't mind libopentelemetry-cpp-header-only either.

I think we still want the CMake metadata so that downstream packages can still `find_package(OpenTelemetry)` and get the headers.

**Comment by h-vetinari** [comment]
> I think we still want the CMake metadata so that downstream packages can still `find_package(OpenTelemetry)` and get the headers.

I think it would be fine to sacrifice that. Adding another include directory is trivial, whereas CMake metadata contains a lot more metadata than just headers.

**Comment by lidavidm** [comment]
Right, but then that means projects that depend on opentelemetry will have to adjust builds to work both against the conda-forge package and a 'regular' installation? I'd like to avoid that if possible, though I suppose that's just the nature of dependencies in C++.

**Comment by h-vetinari** [comment]
> Right, but then that means projects that depend on opentelemetry will have to adjust builds to work both against the conda-forge package and a 'regular' installation?

Hm, I tend to "live" in a world where upstream doesn't care about conda-forge (or barely), so to me it's completely fine if upstream continues working only with `find_package(OpenTelemetry)`, and conda-forge deals with the consequences of its packaging decisions.

I just think that clobbering files is a bad packaging smell, not least because we'll constantly spam every user with warnings about that (but also in general).

**Comment by xhochy** [comment]
> I just think that clobbering files is a bad packaging smell, not least because we'll constantly spam every user with warnings about that (but also in general).

I would strongly support removing any clobbering as this can easily cause issues if you start to patch some of the files in only one feedstock and you cannot rely that the patch is correctly installed in the clobbering situation.

**Comment by xhochy** [comment]
Wouldn't we get around the clobbering problems if this package here depended on `libopentelemetry-cpp-api` in host and run?

**Comment by lidavidm** [comment]
> > Right, but then that means projects that depend on opentelemetry will have to adjust builds to work both against the conda-forge package and a 'regular' installation?
> 
> Hm, I tend to "live" in a world where upstream doesn't care about conda-forge (or barely), so to me it's completely fine if upstream continues working only with `find_package(OpenTelemetry)`, and conda-forge deals with the consequences of its packaging decisions.
> 
> I just think that clobbering files is a bad packaging smell, not least because we'll constantly spam every user with warnings about that (but also in general).

Ok! I'll defer to you and Uwe. I'll give the PR a review in a little bit.

**Comment by lidavidm** [comment]
Merged (and I didn't squash this time!)

**Comment by h-vetinari** [comment]
Opened a PR for the rename here: https://github.com/conda-forge/cpp-opentelemetry-sdk-feedstock/pull/46

After this I think we should be good to add it to the global pinning.
